Abha_Rana/Tic-Tac-Toe/assets/scripts/ttt.js

```diff
@@ -0,0 +1,112 @@
+ var N_SIZE = 3,
+  EMPTY = "&nbsp;",
+  boxes = [],
+  turn = "X",
+  score,
+  moves;
+
+/**
+ * Initializes the Tic Tac Toe board and starts the game.
+ */
+function init() {
+  var board = document.createElement("table");
+  board.setAttribute("border", 1);
+  board.setAttribute("cellspacing", 0);
+
+  var identifier = 1;
+  for (var i = 0; i < N_SIZE; i++) {
+    var row = document.createElement("tr");
+    board.appendChild(row);
+    for (var j = 0; j < N_SIZE; j++) {
+      var cell = document.createElement("td");
+      cell.setAttribute("height", 120);
+      cell.setAttribute("width", 120);
+      cell.setAttribute("align", "center");
+      cell.setAttribute("valign", "center");
+      cell.classList.add("col" + j, "row" + i);
+      if (i == j) {
+        cell.classList.add("diagonal0");
+      }
+      if (j == N_SIZE - i - 1) {
+        cell.classList.add("diagonal1");
+      }
+      cell.identifier = identifier;
+      cell.addEventListener("click", set);
+      row.appendChild(cell);
+      boxes.push(cell);
+      identifier += identifier;
+    }
+  }
+
+  document.getElementById("tictactoe").appendChild(board);
+  startNewGame();
+}
+
+/**
+ * to start New game
+ */
+function startNewGame() {
+  score = {
+    X: 0,
+    O: 0
+  };
+  moves = 0;
+  turn = "X";
+  boxes.forEach(function(square) {
+    square.innerHTML = EMPTY;
+  });
+}
+
+/**
+ * Check if a win or not
+ */
+function win(clicked) {
+  // Get all cell classes
+  var memberOf = clicked.className.split(/\s+/);
+  for (var i =0; i < memberOf.length; i++) {
+    var testClass = "." + memberOf[i];
+    var items = contains("#tictactoe " + testClass, turn);
+    // winning condition: turn == N_SIZE
+    if (items.length == N_SIZE) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * Helper function to check if NodeList from selector has a particular text
+ */
+function contains(selector, text)
+  var elements = document.querySelectorAll(selector);
+  return [].filter.call(elements, function(element) {
```

Comment:
why can't we call `filter` directly on `elements`?

Reply:
filter.call -Uses a blank array in order to call the "filter" method
Tells the filter to use list as the array to filter, and uses function(elements) to do the filtering, where  "elements" is the item in the list being tested
If the item has a style attribute that is empty, i.e. has no style applied, it returns true.